virtual_rainforest/core/axes.py

```diff
@@ -1,323 +1,384 @@
     """
 
-    # Check the axis exists
-    if axis not in AXIS_VALIDATORS:
-        raise ValueError(f"Unknown core axis: {axis}")
+    # Get the validators applying to each axis
+    for axis in AXIS_VALIDATORS:
 
-    # Identify the correct validator routine from the data array signature
-    da_dims = set(darray.dims)
-    da_coords = set(darray.coords.variables)
+        validators: list[Type[AxisValidator]] = AXIS_VALIDATORS[axis]
 
-    # Loop over the available validators loooking for a congruent signature
-    # - dims _cannot_ be empty on a DataArray (they default to dim_N strings) so are
-    #   guaranteed to be non-empty and must then match.
-    # - coords _can_ be empty: they just associate values with indices along the
-    #   dimension. So, the signature should match _specified_ coords but not the
-    #   empty set.
-    for (ld_dims, ld_coords, ld_grid_type), ld_fnm in AXIS_VALIDATORS[axis].items():
+        # Get the set of dim names across all of the validators for this axis
+        validator_dims = set.union(*[v.dim_names for v in validators])
 
-        if (
-            set(ld_dims).issubset(da_dims)
-            and set(ld_coords).issubset(da_coords)
-            and ((data.grid.grid_type in ld_grid_type) or ("any" in ld_grid_type))
-        ):
+        # If the dataarray includes any of those dimension names, one of the validators
+        # for that axis must be able to validate the array, otherwise we can skip
+        # validation on this axis and return the input array.
+        matching_dims = validator_dims.intersection(value.dims)
+        if matching_dims:
 
-            # Retrieve the method associated with the loader signature from the Data
-            # object.
-            return ld_fnm
+            # There should be one and only validator that can validate for this axis.
+            validator_found = [v().can_validate(value, grid) for v in validators]
 
-    return None
+            if not any(validator_found):
+                log_and_raise(
+                    f"DataArray uses '{axis}' axis dimension names but does "
+                    f"not match a validator: {','.join(matching_dims)}",
+                    ValueError,
+                )
 
+            if sum(validator_found) > 1:
+                log_and_raise(
+                    f"Validators on '{axis}' axis not mutually exclusive", RuntimeError
+                )
 
-@register_axis_validator("spatial", (("cell_id",), ("cell_id",), ("any",)))
-def spld_cellid_coord_any(data: Data, darray: DataArray) -> DataArray:
-    """Spatial loader for cell id coordinates onto any grid.
+            # Get the appropriate Validator class and then use it to update the data
+            # array
+            this_validator = validators[validator_found.index(True)]
+            try:
+                value = this_validator().validate(value, grid, **kwargs)
+            except Exception as excep:
+                log_and_raise(str(excep), excep.__class__)
```

Comment:
With this approach, you're running `can_validate` twice and have a too verbose code. If you follow this route, then I'd do the following:


```python
validator_found = [v() for v in validators if v().can_validate(value, grid)]

if len(validator_found) == 0:
    log_and_raise(
        f"DataArray uses '{axis}' axis dimension names but does "
        f"not match a validator: {','.join(matching_dims)}",
        ValueError,
    )

elif len(validator_found) > 1:
    log_and_raise(
        f"Validators on '{axis}' axis not mutually exclusive", RuntimeError
    )

 # Get the appropriate Validator class and then use it to update the data
 # array
 try:
	 value = validators[0].run_validator(value, grid, **kwargs)
 except Exception as excep:
	 log_and_raise(str(excep), excep.__class__)
```

And remove the `validate` class method because it is not necessary.

Reply:
Ha - I did wonder about that 😄 I thought that having that all in one method might be a useful tool at some point, but agree we should keep it simple. I will remove that and merge.

Reply:
Mind that I've also updated how the `validator_found` list is constructed and the conditional statements, making thins, I think, cleaner.